### PR TITLE
OSM support

### PR DESCRIPTION
**Pull request by JasonMWhite**
This is a functional but incomplete reader for OpenStreetMap XML file format. OSM consists of 3 types of objects: Nodes, Ways, and Relations. Nodes == Points, Ways == collections of Nodes, and Relations == collections of (Nodes, Ways, and/or Relations).
http://wiki.openstreetmap.org/wiki/OSM_XML

Complete:
- Nodes -> Points conversion
- Ways -> PolyLine / Polygon conversion

Incomplete:
- Relations -> PolyLine / Polygon(s) conversion
- splitable input format (currently loading whole XML DOM on initial read on one node)

I'm new to Scala (more experienced with PySpark), so I'd appreciate some feedback on my approach up to this point @harsha2010! Also, would it be better to merge this partially functional branch so as to keep the PR smaller, or wait until I complete Relation handling?

### CODE REVIEW

**Comment by harsha2010** [comment]
@JasonMWhite , thanks for the PR. I'll review it and leave some comments here if necessary.
It might be better to wait until Relation handling is available to merge this in, so that it is self contained.

File: src/main/scala/magellan/OsmFileRelation.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+      val reversedPoints = currentPoints.reverse
+      if (wayShapeIsArea(reversedPoints.head, currentPoints.head, tags)) {
+        new Polygon(IndexedSeq(), reversedPoints.toIndexedSeq)
+      } else {
+        new PolyLine(IndexedSeq(), reversedPoints.toIndexedSeq)
+      }
+    }
+  }
+  
+  private def createWayShapes(i: Iterator[(WayKey, (Point, Map[String, String]))], currentId: String, currentPoints: List[Point], currentTags: Map[String, String])
```
Comment (harsha2010) [inline src/main/scala/magellan/OsmFileRelation.scala]: wraparound, can we keep line length 100 characters atmost? similar to Spark's codestyle. https://github.com/databricks/scala-style-guide#linelength

File: src/main/scala/magellan/io/OsmShape.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+case class OsmKey(val shapeType: String, val id: String) extends Serializable { }
+
+abstract class OsmShape(val id: String, val tags: Map[String, String]) extends Serializable { }
+
+case class OsmNode(override val id: String, val lat: Double, val lon: Double, override val tags: Map[String, String]) extends OsmShape(id, tags) {
+  def point: Point = Point(lon, lat)
+}
+
+case class OsmWay(override val id: String, val nodeIds: Seq[String], override val tags: Map[String, String]) extends OsmShape(id, tags) { }
```
Comment (harsha2010) [inline src/main/scala/magellan/io/OsmShape.scala]: line length <= 100., follow spark style guide here: https://github.com/databricks/scala-style-guide#linelength

File: src/test/scala/magellan/OsmSuite.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package magellan
+
```
Comment (harsha2010) [inline src/test/scala/magellan/OsmSuite.scala]: I like these unit tests, could we also add one unit test where an OSM format file is parsed and read ? I don't know much about this format, but maybe mapzen.com has a file that we can use : ideally a fairly small file that we can check in (there are similar small files in src/test/resources for other formats)

**Comment by harsha2010** [comment]
@JasonMWhite , I have completed the review. Let me know when you are ready for me to review and pull in this patch.
We can separate out the optimization(splitting XML) and the polygon collection/ polyline collection implementation into separate PRs to keep each fairly self contained.

**Comment by JasonMWhite** [comment]
Thanks @harsha2010 , I'll get on this probably tonight.

File: src/test/scala/magellan/OsmSuite.scala (hunk 3 above)
Comment (JasonMWhite) [inline src/test/scala/magellan/OsmSuite.scala]: Done

**Comment by JasonMWhite** [comment]
I think this is ready to go. This handles:
- [x] Node -> Point
- [x] Way -> Linestring, Simple Polygons (no holes)
- [ ] Relation -> Polygons (potentially with holes), MultiLinestring, MultiPolygon, MultiPoint (not yet supported)

Also to come in a future PR: file splitting

**Comment by harsha2010** [comment]
@JasonMWhite thanks, this looks good to merge

**Pull request merged**